WordPress/gutenberg
# Components: refactor `Notice` to pass `exhaustive-deps`

## What?
Refactor the Notice component to ignore the `exhaustive-deps` eslint rule. This specifically impacts the `*.native` version of the components

## Why?
Part of the effort in https://github.com/WordPress/gutenberg/pull/41166 to apply `exhuastive-deps` to the Components package

## How?

In `index.native.js`:

Because `startAnimation` is not called anywhere else, its code can be moved directly into the `useEffect` call. This eliminates the need for `startAnimation` as a dependency, but it brings dependencies of its own:
- `animationValue` is actually a `Ref` object, so it should remain stable and safe to list as a dependency
- `onHide` would cause endless re-renders, so it gets wrapped in its own `useCallback` to ensure referencial equality.
- `id` is needed as a dependency for `onHide`'s new `useCallback`. This should be a unique value to the current notice, and not something that should change (though I lost the thread on tracking it. It looks like it's coming from the `@wordpress/notices` store.)
- `onNoticeHidden` is a dependency of both the original `useEffect` and the new `useCallback`. It's a prop passed in from `list.native.js`, and would cause unwanted re-renders, so more changes were needed there.

In `list.native.js`:
`onNoticeHidden` was is being passed the `onRemove` method. To prevent endless re-renders do an unstable reference there, we can wrap `onRemove` in its own `useCallback`. The only dependency here would be the `removeNotice` method, which should be always be a stable reference because it's a dispatch function.

Because all of these changes are in the native components, I lack a good means for testing... @geriux, would you be able to take a look at these changes?

## Testing Instructions

1. From your local Gutenberg directory, run `npx eslint --rule 'react-hooks/exhaustive-deps: warn' packages/components/src/notice`
2. Confirm that the linter returns no errors
